extract/test/test_bbox_merge.py

```diff
@@ -0,0 +1,26 @@
+from extract.bbox import BBox
+from extract.bbox_merge import bbox_merge
+from extract.coordinate import Coordinate
+
+
+def test_bbox_merge():
+    bboxes = [
```

Comment:
so um. this might be a good time to talk about the difference between a list and a tuple..

````python
>>> bboxes = [((1,2),(1,4)),((1,2),(3,4))]
>>> type(bboxes)
<class 'list'>
>>> bboxes = ((1,2),(1,4)),((1,2),(3,4))
>>> type(bboxes)
<class 'tuple'>
>>> 
````

They're similar - both tuples and lists are sequence types. The main difference is that lists are mutable, tuples are immutable. https://docs.python.org/3.7/library/stdtypes.html#sequence-types-list-tuple-range has more:

> Lists are mutable sequences, typically used to store collections of homogeneous items (where the precise degree of similarity will vary by application).
> 
> Lists may be constructed in several ways:
> - Using a pair of square brackets to denote the empty list: `[]`
> - Using square brackets, separating items with commas: `[a]`, `[a, b, c]`
> - Using a list comprehension: `[x for x in iterable]`
> - Using the type constructor: `list()` or `list(iterable)`

> Tuples are immutable sequences, typically used to store collections of heterogeneous data (such as the 2-tuples produced by the enumerate() built-in). Tuples are also used for cases where an immutable sequence of homogeneous data is needed (such as allowing storage in a set or dict instance).
> 
> Tuples may be constructed in a number of ways:
> 
> - Using a pair of parentheses to denote the empty tuple: `()`
> - Using a trailing comma for a singleton tuple: `a`, or `(a,)`
> - Separating items with commas: `a, b, c` or `(a, b, c)`
> - Using the tuple() built-in: `tuple()` or `tuple(iterable)`

Side note: it's not really clear here, but the parentheses don't create a tuple (except for the special case of the empty tuple). It's the *commas* that make a tuple, and *you don't need the parens*. The parens are actually a bit of syntactic fluff that tell the compiler "The stuff inside here should be considered as a single bit of stuff". You can use them to create long strings without making long lines:
````python
>>> m = ("woof"
...    "and stuff"
...    "and things")
>>>    
>>> print(m)
woofand stuffand things
>>> 
````
as long as you avoid the commas that make a tuple:

````python
>>> m = ("woof",
...    "and stuff",
...    "and things")
>>>    
>>> print(m)
('woof', 'and stuff', 'and things')
>>> type(m)
<class 'tuple'>
>>> 
````

Reply:
So what was the point of all this?

I don't think you need to create a list here, I'd just use a tuple.

the bboxes function doesn't care. It wants to get a single object, which is iterable, which it's going to call 'bboxes'. It's not going to mutate that object at all, so an immutable tuple is fine.

On the other hand, I imagine that in real world applications you'll probably use a list rather than a tuple - because you'll start with an empty list and you'll append new bboxes as you find them on the line. so maybe it does make more sense to use a list for testing rather than a tuple... *shrug*

Reply:
https://docs.python.org/3.7/tutorial/datastructures.html#tuples-and-sequences has some good stuff on the differences between tuples and lists and other sequence types.

It also pointed me at https://docs.python.org/3.7/library/collections.html#collections.namedtuple which I wasn't aware of - and the example used there seems quite relevant:

````python
>>> class Point(namedtuple('Point', ['x', 'y'])):
...     __slots__ = ()
...     @property
...     def hypot(self):
...         return (self.x ** 2 + self.y ** 2) ** 0.5
...     def __str__(self):
...         return 'Point: x=%6.3f  y=%6.3f  hypot=%6.3f' % (self.x, self.y, self.hypot)

>>> for p in Point(3, 4), Point(14, 5/7):
...     print(p)
Point: x= 3.000  y= 4.000  hypot= 5.000
Point: x=14.000  y= 0.714  hypot=14.018
````